pkg/catalog/manager/traverse.go

```diff
@@ -54,35 +90,39 @@ func (m *Manager) traverseAndUpdate(helm *helmlib.Helm, commit string, cmt *Cata
 		}
 	}
 
-	entriesWithErrors, entriesToProcess := m.preprocessCatalog(catalog, index.IndexFile.Entries)
-	for chart, metadata := range entriesToProcess {
-		newHelmVersionCommits[chart] = v32.VersionCommits{
-			Value: map[string]string{},
+	catalogTemplateMap, err := m.getTemplateMap(catalogName, templateNamespace)
+	if err != nil {
+		return err
+	}
+
+	// Check if a version has been deleted from a template or if an existing template version has changed.
+	hasChartChanged := func(template *v3.CatalogTemplate, metadata helmlib.ChartVersions) bool {
+		if template == nil || metadata == nil {
+			return false
 		}
-		existingHelmVersionCommits := map[string]string{}
-		if catalog.Status.HelmVersionCommits[chart].Value != nil {
-			existingHelmVersionCommits = catalog.Status.HelmVersionCommits[chart].Value
+		if len(metadata) != len(template.Spec.Versions) {
+			return true
 		}
-		keywords := map[string]struct{}{}
-		// comparing version commit with the previous commit to detect if a template has been changed.
-		hasChanged := false
-		versionNumber := 0
-		for _, version := range metadata {
-			newHelmVersionCommits[chart].Value[version.Version] = version.Digest
-			digest, ok := existingHelmVersionCommits[version.Version]
-			if !ok || digest != version.Digest {
-				hasChanged = true
-			}
-			if ok {
-				versionNumber++
-			}
+
+		metadataVersions := make(map[string]string)
+		for _, metadataVersion := range metadata {
+			metadataVersions[metadataVersion.Version] = metadataVersion.Digest
 		}
-		// if there is a version getting deleted then also set hasChanged to true
-		if versionNumber != len(existingHelmVersionCommits) {
-			hasChanged = true
+		for _, templateVersion := range template.Spec.Versions {
+			// If the digest is not the same between the actual and desired version, or if the version does not
+			// exist in the metadata, then we know that the chart has changed. In addition, we do not have to check
+			// if metadata versions exist because the length of each slice is the same at this point.
+			digest, ok := metadataVersions[templateVersion.Version]
+			if digest != templateVersion.Digest || !ok {
+				return true
+			}
 		}
+		return false
+	}
 
-		if !hasChanged && hasAllUpdates(catalog) {
+	entriesWithErrors, entriesToProcess := m.preprocessCatalog(catalog, index.IndexFile.Entries)
+	for chart, metadata := range entriesToProcess {
+		if !hasChartChanged(catalogTemplateMap[getValidTemplateName(catalogName, chart)], metadata) && hasAllUpdates(catalog) {
```

Comment:
This is going to print logs multiple times. We should save it as a boolean and print log in the end.

Reply:
Also you don't have to call `hasAllUpdates(catalog)` in the loop

Reply:
> This is going to print logs multiple times. We should save it as a boolean and print log in the end.

The anonymous function does not print any logs unless I am missing something.

> you don't have to call hasAllUpdates(catalog) in the loop

The `hasAllUpdates` call was from the previous code. I will call it before the loop begins and store the result in a variable for optimization.

Reply:
https://github.com/rancher/rancher/pull/32992/files/63dd97c65fa60840f208a8edb74cf5c27eba5742#diff-756f710464390ef227414411a01a1f9d7fca7694ffdfd046c22248272412e479R126 this line actually prints it if you turn on debug. We should just print it once

Reply:
Ah good catch. This was from the original code, but I'll move it.

Reply:
Fixed both